### ISSUE TITLE
Have you reproduced the original performance?

### ISSUE THREAD

**Issue by zorrocai** [yxchng/pytorch_face#2]
Hi, yxchng. It is very nice to meet someone working in the same direction.
I just wonder that have you reproduced the original performance reported in the SphereFace paper?

Lots of works have been explored by me to produce the performance on lfw with sphere20 : 

1. The [sphereface](https://github.com/wy1iu/sphereface) has been run with m=4. 
    But I got 98.93%, smaller than the original performance 99.26%.
2. The [sphereface_pytorch](https://github.com/clcarwin/sphereface_pytorch) has been run.
   And the performance is not very good. I find this work has many different settings with the original caffe 
   version.
   1) training data are not clean;
   2) without mirror data augmentation;
   3) learning strategy 20 epochs are bigger than 28000 iterations:
   4) without weights initialization.

Then I have written my own sphereface in pytorch. With 1) 2) 3) fixed, I got 98.18% on LFW with sphere20, still smaller than 99.26%. sphere20 with xavier initialization can not be trained.
Do you have any advice?
Thank you very much.

**Comment by yxchng** [comment]
@zorrocai The best result I got is 99.1% with CASIA-WebFace. There are a few possible explanations:
1) SGDs used in Caffe and Pytorch are different. You can see that my code has two types of SGD - 1) "sgd" and 2) "sutskever_sgd". "sutskever_sgd" is the one used in Caffe. 
2) Different data augmentation
3) Different weight initializations

However, like you said, when I try to train with the original code, I also can't reproduce the result. The best I got with the original Caffe code is also about 99.1%.

I do get 99.6% with another data - MsCeleb.

**Comment by zorrocai** [comment]
Thanks for the reply :). 
After reading your codes, I have some questions. It will be my honour  to have your answers.
1. During training, you have written your own transforms codes. Do the torchvision.transforms have bad 
    influence?
2. During test:
    1) you examine all epochs checkpoints  to choose the best one. Does the last epoch checkpoint is not  
        the best one?
    2) Different from the original caffe version or yours, the best threshold in  [sphereface_pytorch](https://github.com/clcarwin/sphereface_pytorch)  is the last max accuracy threshold. Do you think it effects?

**Comment by yxchng** [comment]
1. Possibly. I am not sure too because I have not compared the differences. I write my own transforms because I like to use OpenCV.
2. 1) The checkpoints are saved based on best loss. However best loss does not equate to best accuracy. Therefore, if you are looking to maximize the score on certain test datasets, you should check all models.
    2) I am not sure what you are asking in this question. The test code follows the test methods used in the original author's Caffe repo, https://github.com/wy1iu/sphereface.

Have you tried training with this code?

**Comment by zorrocai** [comment]
I have not trained with your code.
As for 2 2). I mean the difference between those two functions:
![image](https://user-images.githubusercontent.com/24424749/49588351-71da8f80-f9a1-11e8-89a1-e3243a4c80cf.png)

**Comment by yxchng** [comment]
I am not sure. You have to test yourself.

**Comment by zorrocai** [comment]
Thanks, I have tested it. 
find_best_threshold may be better than bestThreshold.

**Comment by yxchng** [comment]
Great! Hopefully you can inform me if you manage to reproduce the accuracy with PyTorch code.

**Comment by zorrocai** [comment]
Thanks for your advice. I will keep yo in loop if I made some progress.
And one more thing  I think should be noticed is the training epochs.
orginal setting is 28000 iterations with 128 batch size on 4 GPUs.
So the corresponding epochs on 1 GPU should be:
128x4x28000/494414 = 28.995
494414 is the number of images in CASIA.
Considering the missing images that is unable to be detected, I think epochs should be 30.
step values should be 17 and 26, corresponding to  [16000, 24000].
What is your opinion?

**Comment by yxchng** [comment]
I think this may possibly improve the performance. However, note that the optimization process differs in the distributed and single GPU settings, which means you can't just deduce the corresponding number of iterations and expect them to give the same result. 

And yes, this is another big factor that I forgot to mention. Training with 4 GPUs is likely to give better result. I only have one so I can't verify.

**Comment by zorrocai** [comment]
Hi, @yxchng . I have done many experiments on sphereface4. And finally, I got 97.77 on LFW with sphereface4. Here is my tips:
duirng training:
1.use torchvision.transforms
2.trained 30 epochs, step value are 15,23
3.SGD
4.batch size is 512 on 4 GPUs
5.Kaiming uniform initialization

during test:
use your lfw evaluation codes, but without your transforms.
For transforms, just minus 127.5 and divide 128.0 then turn numpy array to tensor.
Do you have any advice to make more 0.43%, since the original percentage is 98.2.

**Comment by yxchng** [comment]
Do you mean sphereface64? You wrote 'sphereface4'. So far, I have only been experimenting with sphereface20.

It is interesting that you only got 97.77. Which dataset did you use to train? I used CASIA-WebFace, which is what the author of the paper used to train. With the default settings of this code, you should get at least 99%. I am not sure where you got the figure 98.2%. According to the paper, you should get 99.42% (an improvement from 98.2%).

![screenshot from 2018-12-18 11-07-07](https://user-images.githubusercontent.com/10518587/50129818-b5909b80-02b5-11e9-8ed6-de199bf3bd8f.png)


For my evaluation code, the transforms to be used depend on the transforms used in your training process. Since you use different transforms, you should use your own. Anyway, my transform is also minus 127.5 and divide 128.0, same as yours.

**Comment by zorrocai** [comment]
I mean 4 layers network, the 98.2 in Figure 6
![image](https://user-images.githubusercontent.com/24424749/50136974-a4a35280-02d4-11e9-9539-de4fdd6a9bb2.png)
